# 4.5 Exercise follow along

## Importing Libraries

In [45]:
import pandas as pd
import numpy as np
import os

## Importing Datasets

In [46]:
#set path variable
path = r'C:\Users\HP\Documents\CareerFoundry\Data Immersion\Achievement 4\05-2023 Instacart Basket Analysis'

In [47]:
path

'C:\\Users\\HP\\Documents\\CareerFoundry\\Data Immersion\\Achievement 4\\05-2023 Instacart Basket Analysis'

In [48]:
#import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [49]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [50]:
#import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [51]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 7 columns]

## Mixed Data Types

In [52]:
#create a test dataframe 

df_test = pd.DataFrame()

In [53]:
#Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [54]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [55]:
#Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [56]:
#replace column data type

df_test['mix'] = df_test['mix'].astype('str')

In [57]:
df_test.dtypes

mix    object
dtype: object

## Missing Data Values

In [58]:
#Check for number of missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [59]:
#Create a subset with only those missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [60]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [61]:
#See dimnensions of original table to compare to dimensions of table with only missing values

df_prods.shape

(49693, 5)

In [62]:
#Create a new dataframe without the rows with missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [63]:
df_prods_clean.shape

(49677, 5)

## Duplicated Values

In [64]:
# Create subset of only rows that are duplicated

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [65]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [66]:
#Check dimensions of original dataframe

df_prods_clean.shape

(49677, 5)

In [67]:
#Drop duplicate values in the dataframe

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [68]:
#Check dimensions to make sure rows were dropped properly

df_prods_clean_no_dups.shape

(49672, 5)

## Exporting dataframe

In [69]:
#Export df_prods as a clean file

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.2

## Step 2: Run df.describe() on df_ords.  Write if anything looks off about the data or should be investigated further.

In [70]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

### The max and min values for each column don't seem to stand out.
### The only thing that might be off is the count for "days_since_last_order" since its count doesn't match the others.

## Step 3: Check for mixed data in df_ords. 

In [71]:
#Checked for mixed data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, 'mixed')
  else:
    print(col, 'consistent')

Unnamed: 0 consistent
order_id consistent
user_id consistent
order_number consistent
orders_day_of_week consistent
order_hour_of_day consistent
days_since_last_order consistent


In [72]:
df_ords.dtypes

Unnamed: 0                 int64
order_id                   int64
user_id                    int64
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

## Step 4. If there are any columns with mixed data types, change them.

### There were no mixed data type columns found.

## Step 5. Run a check for missing values in df_ords. Write a possible explanation for any missing values.

In [73]:
#Check for missing values

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [74]:
#View missing values

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [75]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

### As mentioned in step 3 above, as seen by the count disparity in the "days_since_last_order" and the other columns, there are missing values in the "days_since_last_order" column. These values may be missing since during a customer's first registered order, there shouldn't be any value in the "days_since_last_order" column.

## Step 6. Address the missing values using an appropriate method. 

### I chose to do nothing with the missing values. Imputing a value of 0 would make it more difficult to differentiate between a customer's first order date or just another day where the customer needed to buy something. Removing all rows with the missing values would delete data that could be useful (206209 rows of data). Creating a column that flags missing values could be done but would introduce another column that would need to be analyzed. The data for that column should not be affected too much when looking at statistics. 

## Step 7: Run a check for duplicate values in df_ords

In [76]:
#Checking for duplicate values

df_ords_dups = df_ords[df_ords.duplicated()]

In [77]:
#Viewing duplicate values

df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [78]:
df_ords_dups.shape

(0, 7)

### There are no duplicated values in df_ords.

## Step 8: Address the duplicates with an appropriate method and explain why.

### Since no duplicates were found, there is nothing to address. However, if there were duplicates, they would need to be removed using the 'drop_duplicates()' function. Duplicate values would throw off analyses.

## Step 9: Export df_ords as a .csv file in Prepared Data Folder.

In [79]:
#Exporting df_ords as a new file

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

### I accidentally already exported df_prods earlier while going along with the exercise. It should be located just above step 2 for the Task.